Add Valencia FixConservatives

carmaza opened the pull request:
## Proposed changes


### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add FixConservatives for Valencia

carmaza commented:
Travis will complain about the dox because there is a duplicated class (`FixConservatives`) that can't be resolved (conflict with `ValenciaDivClean`'s). Apparently, wrapping the `ValenciaDivClean` class in the `grmhd::ValenciaDivClean` namespace won't help. Any suggestions for solving this are appreciated.

nilsdeppe commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.hpp:
```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+#include <limits>
```
`cstddef`

nilsdeppe commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.hpp:
```diff
@@ -0,0 +1,127 @@
+  struct SafetyFactorForS {
+    using type = double;
+    static constexpr OptionString help = {
+        "Safety factor for momentum density bound."};
```
This has an upper bound of 1, and lower bound of `epsilon()`. It should also be documented that `1-SafetyFactorForS` is used, since this informs how small is reasonable.

nilsdeppe commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.cpp:
```diff
@@ -0,0 +1,137 @@
+      rest_mass_density_times_lorentz_factor_cutoff_) {
+    PARSE_ERROR(context,
+                "The minimum value of the rest mass density times the Lorentz "
+                "factor  must be less than or equal to the cutoff value of D. "
```
`factor  must` -> `factor must` (remove extra space)

I'd end this with a newline, `D.\n` so that the next part is printed on a separate line. That'll help with readability of the error.

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,101 @@
+SPECTRE_TEST_CASE("Unit.RelativisticEuler.Valencia.FixConservatives",
+                  "[Unit][RelativisticEuler]") {
+  test_fix_conservatives(RelativisticEuler::Valencia::FixConservatives<1>{
+      1.e-12, 1.2e-12, 1.0e-11});
```
you inconsistently use `1.0` and `1.`. Please be consistent within each file.

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,101 @@
+    tilde_s.get(i) = DataVector{1.25, -1.56, 4.32, 5.0, 0.0};
+  }
+  // Use identity matrix as metric for simplicity
+  auto sqrt_det_spatial_metric =
```
I think it makes more sense to specify the inverse spatial metric and compute the determinant from it than the other way around.

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,101 @@
+      for (size_t i = 0; i < Dim; ++i) {
+        expected_tilde_s_after_fixing.get(i)[s] *= std::min(
+            1.0,
+            sqrt((1 - safety_factor_for_s_tilde) * tilde_s_squared_max /
```
`1.0`

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,101 @@
+  CHECK_ITERABLE_APPROX(tilde_tau, expected_tilde_tau_after_fixing);
+  CHECK_ITERABLE_APPROX(tilde_s, expected_tilde_s_after_fixing);
+
+  test_serialization(variable_fixer);
```
I would recommend also running the above checks on the deserialized variable fixer. We've had issues with `operator==` being implemented incorrectly and then this check passing.

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,101 @@
+
+  test_serialization(variable_fixer);
+
+  const auto fixer_from_options =
```
Could you also run the above tests on the created fixer?

carmaza committed:
Add missing cond in ValenciaDivClean FixConservatives

carmaza committed:
Wrap ValenciaDivClean FixConservatives in system namespace

carmaza commented:
Rebased on develop. I added a commit fixing a missing `\cond` in `ValenciaDivClean`'s `FixConservatives`, which fixes the ambiguity in the documentation, and I added another commit that moves `ValenciaDivClean`'s `FixConservatives` to the `grmhd::ValenciaDivClean` namespace, in consistency with what I did for `Valencia`'s. These new commits go before the original commit adding `FixConservatives` for `Valencia`, so each commit passes all builds independently.

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -78,7 +82,14 @@ void test_fix_conservatives(
+template <size_t Dim>
+void test_fixer() noexcept {
+  RelativisticEuler::Valencia::FixConservatives<Dim> variable_fixer(
+      1.e-12, 1.2e-12, 1.0e-11);
```
`1.` -> `1.0`

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -87,15 +98,19 @@ void test_fix_conservatives(
           "  CutoffD: 1.2e-12\n"
           "  SafetyFactorForS: 1.0e-11");
   CHECK(variable_fixer == fixer_from_options);
```
Please also add a `test_fix_conservatives<Dim>(fixer_from_options);`

nilsdeppe reviewed (changes requested):
I'm happy for you to squash, including these last two minor things

carmaza commented:
Squashed and rebased on develop.

nilsdeppe reviewed (dismissed):
ClangTidy error is the silly thing inside boost where the root finders don't initialize some variable and ClangTidy gets confused...

kidder commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,117 @@
+  // assuming that determinant_and_inverse is correctly implemented,
+  // the metric should be its inverse, and its determinant should equal its sqrt
+  auto sqrt_det_spatial_metric =
+      determinant_and_inverse(inv_spatial_metric).first;
```
just call `determinant`

kidder commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,117 @@
+        (get(expected_tilde_tau_after_fixing)[s] +
+         2.0 * get(expected_tilde_d_after_fixing)[s]);
+    if (get(tilde_s_squared)[s] >
+        (1 - safety_factor_for_s_tilde) * tilde_s_squared_max) {
```
1.0

kidder reviewed (changes requested):
(reviewed with @wthrowe ) you can squash

kidder commented:
clang tidy error is known issue with boost, so we will ignore it

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.cpp:
```diff
@@ -0,0 +1,139 @@
+  if (minimum_rest_mass_density_times_lorentz_factor_ >
+      rest_mass_density_times_lorentz_factor_cutoff_) {
+    PARSE_ERROR(context,
+                "The minimum value of the rest mass density times the Lorentz "
```
The first sentence would be easier to read if rephrased as: "The cutoff value of D (D = rest mass density * Lorentz factor) must be not be below the minimum value of D."

carmaza replied:
Done.

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.cpp:
```diff
@@ -0,0 +1,139 @@
+      temp_buffer(number_of_points);
+
+  // rest mass density times Lorentz factor
+  DataVector& rho_w = get(get<::Tags::TempScalar<0>>(temp_buffer));
```
This is the variable D, and you use D in the documentation. Why not use the consistent `d` in the code?

carmaza replied:
I don't think we like single-character identifiers. However, I added a comment clarifying that this is indeed D.

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.hpp:
```diff
@@ -0,0 +1,134 @@
+ * \brief Fix conservative variables using the method proposed in F. Foucart's
+ * PhD thesis (Cornell)
+ *
+ * Fix the conservative variables according to the following recipe:
```
I suggest being more explicit that this is a pointwise operation. Perhaps "Fix the conservative variables at each grid point according to..."

carmaza replied:
Done.

fmahebert commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,116 @@
+  }
+  // assuming that determinant is correctly implemented,
+  // the determinant should equal its square root
+  auto sqrt_det_spatial_metric = determinant(inv_spatial_metric);
```
Is my understanding correct? Doesn't this only work because you've used an identity matrix for the metric, giving a determinant of 1? So... why not just write `const auto sqrt_det_spatial_metric = Scalar<double>{{{1.0}}};`?

carmaza replied:
Your understanding is correct. I'm only using `determinant` for a bit of extra consistency. Not a big deal in a test file, I guess.

fmahebert commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_FixConservatives.cpp:
```diff
@@ -0,0 +1,116 @@
+      DataVector{0.0, 0.7, 0.1, 5.5, 0.0}};
+
+  auto expected_tilde_s_after_fixing = tilde_s;
+  const double safety_factor_for_s_tilde = 1.0e-11;
```
Why are you using "s_tilde" here instead of "tilde_s" like everywhere else? This is surprising when reading the code.

carmaza replied:
Good catch. Done.

fmahebert reviewed (commented):
This is a little late, so consider non-binding.

carmaza commented:
Addressed @fmahebert 's comments, squashed immediately, then rebased on develop in order to fix a small conflict.

nilsdeppe commented on src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.cpp:
```diff
@@ -0,0 +1,139 @@
+    PARSE_ERROR(
+        context,
+        "The cutoff value of D (D = rest mass density * Lorentz factor) must "
+        "be not be below the minimum value of D.\nValues given: D_min = "
```
`be not be` -> `not be`

carmaza replied:
RIP

fmahebert replied:
RIP me, sorry 😞

kidder commented:
ignoring travis problems
